lib/service_base.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+class ServiceBase
+  class ServiceFailureError < StandardError
+    attr_reader :result
+
+    def initialize(result)
+      @result = result
+    end
+  end
+
+  attr_reader :guardian
+
+  include HasErrors
+
+  FAIL_TYPES =
+    Enum.new(permission: 1, validation: 2, record_not_found: 3, unexpected: 4, uncaught: 99)
+
+  def fail_permissions!(messages)
+    raise ServiceFailureError.new(
+            ServiceResult.failed(errors: messages, failure_type: FAIL_TYPES[:permission]),
+          )
+  end
+
+  def fail_validation!(messages)
+    raise ServiceFailureError.new(
+            ServiceResult.failed(errors: messages, failure_type: FAIL_TYPES[:validation]),
+          )
+  end
+
+  def fail_record_not_found!(message)
+    raise ServiceFailureError.new(
+            ServiceResult.failed(errors: messages, failure_type: FAIL_TYPES[:record_not_found]),
+          )
+  end
+
+  def fail_unexpected!(message)
+    raise ServiceFailureError.new(
+            ServiceResult.failed(errors: messages, failure_type: FAIL_TYPES[:unexpected]),
+          )
+  end
+
+  def fail_uncaught!(raised_error)
+    ServiceResult.failed(
+      errors: uncaught_error_message,
+      failure_type: FAIL_TYPES[:record_not_found],
+      raised_error: raised_error,
+    )
+  end
+
+  def execute_service_call
```

Comment:
Wrapping everything in these `execute_service_call` calls lets us handle errors much more gracefully especially the uncaught ones, and makes sure we always are returning a `ServiceResult`